### PR TITLE
fix(agw): Fixed memory leak found on network initiated detach procedure

### PR DESCRIPTION
**Pull request by rsarwad**
fix(agw): Fixed memory leak found on network initiated detach procedure

## Summary
For network initiated detach procedure, mme starts 3422 timer and allocates memory to store the required information. As part of handling detach, mme needs to clear the in-memory contexts, As part of clearing in-memory contexts, memory allocated for 3422 timer was not freed.
Now added fix to clear memory allocated for 3422 timer while clearing in-memory contexts.
## Test Plan
Executed s1ap sanity test suite

With fixes, there are no memory leaks found.
Without fix, memory leaks were found

### CODE REVIEW

**Review by ulaskozat — APPROVED** [review]
LGTM. Can you also backport this to v1.6?

**Comment by ulaskozat** [comment]
@rsarwad this looks like failing a unit test.

**Comment by rsarwad** [comment]
> LGTM. Can you also backport this to v1.6?

This leak observed after porting timers to zmq. v1.6 doesn't include zmq timer. Backporting this PR is not required.

**Comment by ulaskozat** [comment]
> > LGTM. Can you also backport this to v1.6?
> 
> This leak observed after porting timers to zmq. v1.6 doesn't include zmq timer. Backporting this PR is not required.

How does migration to zmq affect this?

**Comment by ulaskozat** [comment]
@rsarwad any plans to revisit this?

**Comment by rsarwad** [comment]
> @rsarwad any plans to revisit this?

Hi @ulaskozat, I was working on adding unit test cases for dedicated bearer creation and deletion. Planning spare some time today.

**Comment by rsarwad** [comment]
> > > LGTM. Can you also backport this to v1.6?
> > 
> > 
> > This leak observed after porting timers to zmq. v1.6 doesn't include zmq timer. Backporting this PR is not required.
> 
> How does migration to zmq affect this?

Yes @ulaskozat, my analysis is wrong. But let me find the cause why the leak is not shown on branch v1.6

File: lte/gateway/c/core/oai/tasks/nas/emm/EmmCommon.c [hunk 1]
```diff
@@ -501,8 +502,12 @@ void create_new_attach_info(
   OAILOG_FUNC_IN(LOG_NAS_EMM);
   emm_context_p->new_attach_info = calloc(1, sizeof(new_attach_info_t));
   emm_context_p->new_attach_info->mme_ue_s1ap_id = mme_ue_s1ap_id;
-  emm_context_p->new_attach_info->ies            = ies;
-  emm_context_p->new_attach_info->is_mm_ctx_new  = is_mm_ctx_new;
+  emm_context_p->new_attach_info->ies =
+      (emm_attach_request_ies_t*) calloc(1, sizeof(emm_attach_request_ies_t));
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas/emm/EmmCommon.c]: `ies` is a stolen reference. you do not need to calloc.

**Review by ulaskozat — CHANGES_REQUESTED** [review]
I recently fixed a memory leak. please double check with that.

**Comment by ulaskozat** [comment]
which test are you using for memory leak?

File: lte/gateway/c/core/oai/tasks/nas/emm/Attach.c [hunk 2]
```diff
@@ -2597,10 +2597,16 @@ void proc_new_attach_req(
...
+  new_attach_info_t attach_info              = {0};
+  struct emm_attach_request_ies_s attach_ies = {0};
   memcpy(
       &attach_info, ue_context_p->emm_context.new_attach_info,
       sizeof(new_attach_info_t));
+  memcpy(
+      &attach_ies, ue_context_p->emm_context.new_attach_info->ies,
+      sizeof(emm_attach_request_ies_t));
+  attach_info.ies = &attach_ies;
+  free_wrapper((void**) &ue_context_p->emm_context.new_attach_info->ies);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas/emm/Attach.c]: I don't think this is needed...

Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/nas/emm/Attach.c]: As per the discussion, in this PR retaining fixes only network initiated detach procedure

**Review by ulaskozat — APPROVED** [review]
LGTM

**Comment by rsarwad** [comment]
> which test are you using for memory leak?

The leak was observed for TC,  test_attach_nw_initiated_detach_with_mme_restart.py

File: lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.c [hunk 3]
```diff
@@ -653,6 +653,9 @@ void free_emm_ctx_memory(
   if (!ctxt) {
     return;
   }
+  if (ctxt->t3422_arg) {
+    free_wrapper((void**) &ctxt->t3422_arg);
+  }
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.c:658]: isn't this handled as part of `nas_delete_all_emm_procedures`?

Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.c:658]: There is no emm procedure created for detach procedure, so condition, https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/nas/nas_procedures.c#L563 fails

Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/nas/emm/emm_data_ctx.c:658]: the question then is "should it be created". If we do not have a "detach procedure", let's clean it up.

**Pull request merged**